### PR TITLE
[corechecks/snmp] Add User Profiles support

### PR DESCRIPTION
**Pull request by AlexandreYang**
### What does this PR do?

Add User Profiles support

- `profiles` is used for user profiles
- `default_profiles` will contain datadog ootb profiles
- `profiles` (user) have precedence over `default_profiles`
- If the same sysObjectID is used in both `profiles` and `default_profiles`, the one from `profiles` will take precedence. The same sysObjectID cannot be used in the same profile folder as before.
- `extends` will favour profiles from users `profiles` folder

### Motivation

Separate default and user profiles and provide a specific folder where users can add their profiles.

### Describe how to test/QA your changes

1/ Check that all the feature changes described in PR Description works:
- `profiles` (user) have precedence over `default_profiles`
- If the same sysObjectID is used in both `profiles` and `default_profiles`, the one from `profiles` will take precedence. The same sysObjectID cannot be used in the same profile folder as before.
- `extends` will favour profiles from users `profiles` folder

2/ ⚠️ Test that the SNMP integration works as expected when users migrate from previous versions.
For example:
- Make sure that `snmp.d/profiles` (user) does not contain the OOTB profiles anymore
- Make sure that the OOTB are now in `snmp.d/default_profiles` folder
- Make sure that user profiles work as expected after the migration
- Make sure that default profiles work as expected after the migration
- Test as much cases as possible to detect any behaviour that might lead to a breaking change.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: omnibus/config/software/datadog-agent-integrations-py2.rb [hunk 1]
```diff
@@ -432,9 +432,12 @@
         end
 
         # Copy SNMP profiles
-        profiles = "#{check_dir}/datadog_checks/#{check}/data/profiles"
-        if File.exist? profiles
-          copy profiles, "#{check_conf_dir}/"
+        profile_folders = ['profiles', 'user_profiles']
+        profile_folders.each do |profile_folder|
```
Comment (AlexandreYang) [inline omnibus/config/software/datadog-agent-integrations-py2.rb:436]: TODO: Move to separate PR: https://github.com/DataDog/datadog-agent/pull/17631

**Review by buraizu — APPROVED** [review]

File: pkg/collector/corechecks/snmp/internal/checkconfig/config.go [hunk 2]
```diff
@@ -733,8 +733,13 @@ func GetProfileForSysObjectID(profiles profileConfigMap, sysObjectID string) (st
 			if !found {
 				continue
 			}
-			if matchedProfile, ok := tmpSysOidToProfile[oidPattern]; ok {
-				return "", fmt.Errorf("profile %s has the same sysObjectID (%s) as %s", profile, oidPattern, matchedProfile)
+			if prevMatchedProfile, ok := tmpSysOidToProfile[oidPattern]; ok {
+				if profiles[prevMatchedProfile].isUserProfile && !profConfig.isUserProfile {
```
Comment (FlorianVeaux) [inline pkg/collector/corechecks/snmp/internal/checkconfig/config.go:737]: Trying to make sure I read this right, but it seems to me like this will give precedence to _any_ user profile versus our profiles.
Meaning that if they have a simple "base" profile, matching on `1.3.6.1.4.*` all devices will start to use this profile instead.
I thought we wanted to give precedence to custom profiles only when they match on the same sysObjectID

File: pkg/collector/corechecks/snmp/internal/checkconfig/profile.go [hunk 3]
```diff
@@ -101,24 +126,23 @@ func getDefaultProfilesDefinitionFiles() (profileConfigMap, error) {
...
 			if err != nil {
 				log.Warnf("failed to expand profile `%s`: %s", name, err)
 				continue
 			}
-			profile.Definition = *profileDefinition
-			profiles[name] = profile
-		} else {
-			profiles[name] = profile
+			profConfig.Definition = *profDefinition
+			profiles[name] = profConfig
```
Comment (FlorianVeaux) [inline pkg/collector/corechecks/snmp/internal/checkconfig/profile.go:143]:
```suggestion
```
I don't think we need this line now that you've moved it outside of the if block, do we?

File: pkg/collector/corechecks/snmp/internal/checkconfig/config.go (hunk 2 above)
Comment (AlexandreYang) [reply pkg/collector/corechecks/snmp/internal/checkconfig/config.go:737]: The way it's currently implemented, the 1/ merge of `profiles` and `default_profiles` and the 2/ sysObjectID matching are two different steps.

In step 1/ when merging process doesn't rely on the sysObjectID. And the sysObjectID is currently not inherited, only the sysObjectID of the leaf profile is considered. If the user profile override a base profile that `1.3.6.1.4.*`, the user profile will take precedence the sysObjectID of the user profile will be used.
In case the user want to use a different sysObjectID and not override, maybe it's easier to just create a profile with different name and extends the "base" profile with `1.3.6.1.4.*`.

> I thought we wanted to give precedence to custom profiles only when they match on the same sysObjectID

We give precedent to user profile when the same sysObjectID is present multiple profiles in `profiles` and `default_profiles`.

File: pkg/collector/corechecks/snmp/internal/checkconfig/profile.go (hunk 3 above)
Comment (AlexandreYang) [reply pkg/collector/corechecks/snmp/internal/checkconfig/profile.go:143]: Good catch!

File: pkg/collector/corechecks/snmp/internal/checkconfig/config.go (hunk 2 above)
Comment (FlorianVeaux) [reply pkg/collector/corechecks/snmp/internal/checkconfig/config.go:737]: I'm really talking about step 2/ here. My example is let's say a user has a single custom profile `/snmp.d/profiles/my-custom-profile.yaml` with the following:

```yaml
sys_object_ids:
  - 1.*

metrics:
 ...
```

What I understand is that this custom profile will take precedence over all datadog OOTB profiles and I'm wondering if that's the intention

Comment (AlexandreYang) [reply pkg/collector/corechecks/snmp/internal/checkconfig/config.go:737]: > What I understand is that this custom profile will take precedence over all datadog OOTB profiles and I'm wondering if that's the wanted behavior

Hum, that's actually not the case. Added this additional unit test to cover this exact case: https://github.com/DataDog/datadog-agent/pull/17395/commits/1961c671f45f004ee2ff85e688a923a2d68b45ac

(maybe I'm misunderstanding your point)

@FlorianVeaux

Comment (AlexandreYang) [reply pkg/collector/corechecks/snmp/internal/checkconfig/config.go:737]: Note: the variable `tmpSysOidToProfile` is a simple `set` of sysObjectIDs like `{"1.*", "1.2.3"}` and in your example, if `profiles/my-custom-profile.yaml` has sysobjectid: 1.*, and `default_profiles/some_profile.yaml` has "1.2.3".

At this point of the code, `tmpSysOidToProfile` will have both entries.

The resolution is done few lines later with `getMostSpecificOid(matchedOids)`.

Comment (FlorianVeaux) [reply pkg/collector/corechecks/snmp/internal/checkconfig/config.go:737]: Makes sense, thanks for explaining and adding the unit test!

**Review by FlorianVeaux — APPROVED** [review]

**Review by ken-schneider — APPROVED** [review]
